File: backend/app/services/llm_service.py
from langchain_core.messages import HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI

from ..config import settings


def _assert_llm_config() -> None:
    if not settings.llm_api_key:
        raise ValueError("缺少 LLM_API_KEY 或 DASHSCOPE_API_KEY 环境变量")
    if not settings.llm_base_url:
        raise ValueError("缺少 LLM_BASE_URL 环境变量")
    if not settings.llm_model:
        raise ValueError("缺少 LLM_MODEL 环境变量")


def _create_chat_model() -> ChatOpenAI:
    _assert_llm_config()

    # 这里统一走 LangChain 的 ChatOpenAI，并通过 base_url 接阿里云百炼这类
    # OpenAI-compatible 平台。这样后面做 RAG 时，可以继续沿用 LangChain 生态。
    return ChatOpenAI(
        api_key=settings.llm_api_key,
        model=settings.llm_model,
        temperature=settings.llm_temperature,
        max_tokens=settings.llm_max_tokens,
        max_retries=0,
        base_url=settings.llm_base_url,
        streaming=True,
    )


def _build_messages(system_prompt: str, user_prompt: str) -> list:
    return [SystemMessage(content=system_prompt), HumanMessage(content=user_prompt)]


def _normalize_chunk_text(content) -> str:
    # LangChain 在不同模型下返回的 chunk.content 结构不完全一致：
    # 有时是纯字符串，有时是内容块列表。这里统一压成纯文本。
    if isinstance(content, str):
        return content

    if not isinstance(content, list):
        return ""

    parts: list[str] = []
    for item in content:
        if isinstance(item, str):
            parts.append(item)
            continue

        text = getattr(item, "text", None)
        if isinstance(text, str):
            parts.append(text)
            continue

        if isinstance(item, dict) and isinstance(item.get("text"), str):
            parts.append(item["text"])

    return "".join(parts)


def _format_model_error(error: Exception) -> str:
    return (
        "LangChain 模型调用失败: "
        f"baseURL={settings.llm_base_url} "
        f"model={settings.llm_model} "
        f"reason={error}"
    )


def _collect_stream_reply(system_prompt: str, user_prompt: str):
    model = _create_chat_model()
    full_reply = ""

    # 这里直接使用 LangChain 的 stream()。
    # 对只支持流式输出的模型，这条链路比普通 invoke 更稳。
    for chunk in model.stream(_build_messages(system_prompt, user_prompt)):
        text = _normalize_chunk_text(chunk.content)

        if not text:
            continue

        full_reply += text
        yield text

    if not full_reply.strip():
        raise ValueError("模型流式输出为空")


def generate_reply(system_prompt: str, user_prompt: str) -> str:
    try:
        # 即使对外暴露的是“普通回复”接口，内部仍然走流式聚合。
        # 这样流式和非流式接口共用同一套模型调用路径，行为更一致。
        chunks = list(_collect_stream_reply(system_prompt, user_prompt))
        reply = "".join(chunks).strip()
        if not reply:
            raise ValueError("模型流式输出为空")
        return reply
    except Exception as error:
        raise RuntimeError(_format_model_error(error)) from error


def stream_reply(system_prompt: str, user_prompt: str):
    try:
        for chunk in _collect_stream_reply(system_prompt, user_prompt):
            yield chunk
    except Exception as error:
        raise RuntimeError(_format_model_error(error)) from error

File: backend/src/services/llmService.js
import { ChatOpenAI } from '@langchain/openai'
import { HumanMessage, SystemMessage } from '@langchain/core/messages'
import { env } from '../config/env.js'

function assertLlmConfig() {
  if (!env.llmApiKey) {
    throw new Error('缺少 LLM_API_KEY 或 DASHSCOPE_API_KEY 环境变量')
  }

  if (!env.llmBaseUrl) {
    throw new Error('缺少 LLM_BASE_URL 环境变量')
  }

  if (!env.llmModel) {
    throw new Error('缺少 LLM_MODEL 环境变量')
  }
}

function createChatModel() {
  assertLlmConfig()

  // 这里只把 LangChain 限定在“模型调用层”：
  // controller 仍然只管收参和返回，prompt 仍然由 promptService 维护。
  // 这样后续即使继续接 RAG，也只需要在 service 层往下扩展。
  return new ChatOpenAI({
    apiKey: env.llmApiKey,
    model: env.llmModel,
    temperature: env.llmTemperature,
    maxTokens: env.llmMaxTokens,
    maxRetries: 0,
    configuration: {
      baseURL: env.llmBaseUrl,
    },
  })
}

function buildMessages({ systemPrompt, userPrompt }) {
  return [new SystemMessage(systemPrompt), new HumanMessage(userPrompt)]
}

function normalizeChunkText(content) {
  // LangChain 在不同模型下返回的 content 可能是字符串，
  // 也可能是由多个内容块组成的数组。这里统一提取成纯文本，
  // 这样上层 controller 和前端都不用关心底层差异。
  if (typeof content === 'string') {
    return content
  }

  if (!Array.isArray(content)) {
    return ''
  }

  return content
    .map((item) => {
      if (typeof item === 'string') {
        return item
      }

      if (typeof item?.text === 'string') {
        return item.text
      }

      return ''
    })
    .join('')
}

function formatModelError(error) {
  const reason = error instanceof Error ? error.message : 'unknown error'

  return `LangChain 模型调用失败: baseURL=${env.llmBaseUrl} model=${env.llmModel} reason=${reason}`
}

async function collectStreamReply({ systemPrompt, userPrompt, onChunk }) {
  const model = createChatModel()
  const stream = await model.stream(buildMessages({ systemPrompt, userPrompt }))
  let fullReply = ''

  for await (const chunk of stream) {
    const text = normalizeChunkText(chunk.content)

    if (!text) {
      continue
    }

    fullReply += text

    if (typeof onChunk === 'function') {
      onChunk(text)
    }
  }

  if (!fullReply.trim()) {
    throw new Error('模型流式输出为空')
  }

  return fullReply.trim()
}

export async function generateReply({ systemPrompt, userPrompt }) {
  try {
    // 即使是“普通回复”接口，这里也统一走流式聚合。
    // 原因是你当前使用的部分模型只支持流式 HTTP 调用，
    // 这样可以保证非流式接口和流式接口共用同一套稳定路径。
    return await collectStreamReply({ systemPrompt, userPrompt })
  } catch (error) {
    throw new Error(formatModelError(error))
  }
}

export async function streamReply({ systemPrompt, userPrompt, onChunk }) {
  try {
    return await collectStreamReply({ systemPrompt, userPrompt, onChunk })
  } catch (error) {
    throw new Error(formatModelError(error))
  }
}
